kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/slot-updates.k

```diff
@@ -0,0 +1,143 @@
+    // where the masking clears the part of the slot into which VALUE will be stored,
+    // and for VALUE to be stored correctly it first has to be shifted appropriately.
+    // Note that SHIFT and MASK are always concrete.
+    //
+    // We perform this update in several stages:
+    // 1. First, we simplify MASK &Int #asWord( SLOT ), which results in
+    //    ( VALUE *Int SHIFT ) |Int #asWord ( B1 +Bytes ... +Bytes BN ).
+    // 2. Then, we isolate the +Bytes-junct(s) that will be overwritten.
+    // 3. Then, we write the VALUE, possibly splitting the identified +Bytes-junct(s).
+    //
+    // Note that we require additional simplifications to account for the fact that
+    // VALUE and SLOT can also be concrete. In the former case, we need to extract the
+    // SHIFT appropriate, and in the latter case, the slot will appear on the LHS of the |Int.
+
+    // 1. Slot masking using &Int
+    rule [mask-b-and]:
+      MASK:Int &Int SLOT:Int =>
+        #asWord ( #buf ( 32, SLOT ) [ 32 -Int ( #getMaskShiftBytes(MASK) +Int #getMaskWidthBytes(MASK) ) := #buf ( #getMaskWidthBytes(MASK), 0 ) ] )
+        requires #isMask(MASK) andBool #rangeUInt(256, SLOT)
+        [simplification, concrete(MASK), preserves-definedness]
+
+    // 2a. |Int and +Bytes, update to be done in left
+    rule [bor-update-to-left]:
+      A |Int #asWord ( B1 +Bytes B2 ) =>
+        #asWord ( #buf ( 32 -Int lengthBytes(B2), (A /Int (2 ^Int (8 *Int lengthBytes(B2)))) |Int #asWord ( B1 ) ) +Bytes B2 )
+        requires #rangeUInt(256, A) andBool A modInt (2 ^Int (8 *Int lengthBytes(B2))) ==Int 0 andBool lengthBytes(B1 +Bytes B2) <=Int 32
+        [simplification, preserves-definedness]
+
+    // 2b. |Int of +Bytes, update to be done in right
+    rule [bor-update-to-right]:
+      A |Int #asWord ( B1 +Bytes B2 ) =>
+        #asWord ( B1 +Bytes #buf ( lengthBytes(B2), A |Int #asWord ( B2 ) ) )
+        requires 0 <=Int A andBool A <Int 2 ^Int (8 *Int lengthBytes(B2)) andBool lengthBytes(B2) <=Int 32
+        [simplification, preserves-definedness]
+
+    // 3a. Update with explicit shift and symbolic slot
+    rule [bor-update-with-shift]:
+      ( SHIFT *Int X ) |Int Y => #asWord ( #buf( 32 -Int ( log2Int(SHIFT) /Int 8 ), X ) +Bytes #buf( log2Int(SHIFT) /Int 8, Y ) )
+      requires rangeUInt(256, SHIFT) andBool SHIFT ==Int 2 ^Int log2Int(SHIFT) andBool log2Int(SHIFT) modInt 8 ==Int 0
+       andBool 0 <=Int X andBool X <Int 2 ^Int (8 *Int (32 -Int ( log2Int(SHIFT) /Int 8 )))
+       andBool 0 <=Int Y andBool Y <Int SHIFT
+       [simplification, concrete(SHIFT), comm, preserves-definedness]
+
+    // 3b. Buffer cropping
+    rule [buf-asWord-crop]:
+      #buf (W1:Int, #asWord(#buf(W2, X:Int) +Bytes B)) => #buf(W1 -Int lengthBytes(B), X) +Bytes B
+      requires 0 <=Int W1 andBool 0 <=Int W2 andBool lengthBytes(B) <=Int W1 andBool W1 <=Int W2 +Int lengthBytes(B)
+       andBool 0 <=Int X andBool X <Int 2 ^Int (8 *Int (W1 -Int lengthBytes(B)))
+      [simplification]
```

Comment:
I think this one also suffers from moving something inside the `Word` field to outside the `Word` field. Specifically, `B` is chopped on the LHS of the rule, but free on the RHS.

Suppose:

- `lengthBytes(B) == 34`
- `W2 == 2`
- `W1 == 36`
- `X == 3`

Then, the LHS will basically fully chop the `X` off inside the `#asWord`, and will truncate part of `B` to be the rightmost 32 bytes, but the RHS will keep `X` intact at the front, and leave `B` unchopped.

I guess some restrictions you can add:

```
W2 +Int lengthBytes(B) <=Int 32
```

Reply:
We do need a restriction, but I think the one we want is `W1 <=Int 32`, as the resulting buffer is of length `W1`.

With this in place, `X <Int 2 ^Int (8 *Int (W1 -Int lengthBytes(B)))` means that the chop has nothing to chop, is that right?

Reply:
Adding `W1 <=Int 32` for now.

Reply:
I don't think so actually. `W1 <=Int 32` just means that the outer `#buf(...)` has a total length of `32`, but doesn't say anything about the length of the contents of `#asWord`, that can still be arbitrarily wide.

In fact, the above counterexample with just changing `W1 == 32` works again even with that restriction. `W1 == 32` just means that the second argument of `#buf(..., _)` needs to be within 32 bytes, but `#asWord(...)` ensures that it's the case, even if the contents are arbitrarily long, so `B` can still basically be unbounded in length and this lemma will apply, which will result in something of length than 32 bytes becoming something longer than 32 bytes.

When I'm looking to analyze these things, I basically look for values that start in(out) of the `Word` finite-field, and end out(in) the `Word` finite-field. In this case, because `B` starts inside an `#asWord`, it's effecitively constrained to being in the field on the LHS, but on the RHS it shows up outside the field.

Reply:
Hm, but in the counter-example, `lengthBytes(B) <=Int W1` doesn't hold if `W1 ==Int 32` and `lengthBytes(B) == 34`.

The length of the contents of `#asWord` can indeed be arbitrarily wide, but having `0 <=Int X andBool X <Int 2 ^Int (8 *Int (W1 -Int lengthBytes(B)))` together with `W1 <=Int 32` and `lengthBytes(B) <=Int W1` should ensure that there is no overflow.

The goal of the simplification is precisely to trim the leading zeros of the buffer when the overall length (`W2 +Int lengthBytes(B)`) is greater than 32.

Reply:
In any case, I have generalised it to:
```
    // 3b. Buffer cropping
    rule [buf-asWord-crop]:
      #buf (W:Int , #asWord(B:Bytes)) => #range(B, lengthBytes(B) -Int W, W)
      requires 0 <=Int W andBool W <=Int 32 andBool W <Int lengthBytes(B) 
       andBool #asWord ( #range(B, 0, lengthBytes(B) -Int W) ) ==Int 0
      [simplification, concrete(W), preserves-definedness]
```
and added another lemma that trims leading zeros off of a buffer if it's in `#asWord` and longer than 32.